fedimint-api/src/db/mod.rs

```diff
@@ -56,16 +59,83 @@ pub trait DatabaseValue: Sized + SerializableDatabaseValue {
 }
 
 pub type PrefixIter<'a> = Box<dyn Iterator<Item = Result<(Vec<u8>, Vec<u8>)>> + Send + 'a>;
+type Notifications = DashMap<Box<[u8]>, Arc<Notify>>;
 
 #[async_trait]
 pub trait IDatabase: Debug + Send + Sync {
-    async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction;
+    async fn begin_transaction<'a>(&'a self) -> Box<dyn IDatabaseTransaction<'a> + Send + 'a>;
 }
 
-dyn_newtype_define! {
-    /// A handle to a type-erased database implementation
-    #[derive(Clone)]
-    pub Database(Arc<IDatabase>)
+#[derive(Clone, Debug)]
+pub struct Database {
+    inner: Arc<dyn IDatabase>,
+    notifications: Arc<Notifications>,
+}
+
+impl<D: IDatabase + 'static> From<D> for Database {
+    fn from(inner: D) -> Self {
+        Database {
+            inner: Arc::new(inner),
+            notifications: Arc::new(Notifications::new()),
+        }
+    }
+}
+
+impl Database {
+    pub async fn begin_transaction(&self, decoders: ModuleDecoderRegistry) -> DatabaseTransaction {
+        DatabaseTransaction::new(
+            self.inner.begin_transaction().await,
+            decoders,
+            self.notifications.clone(),
+        )
+    }
+
+    pub async fn wait_key<K: DatabaseKey + DatabaseKeyPrefixConst>(
+        &self,
+        key: K,
+        decoders: ModuleDecoderRegistry,
+    ) -> Result<K::Value> {
+        let mut tx = self.inner.begin_transaction().await;
+        let key_bytes = key.to_bytes();
+        let value_bytes = if let Some(value_bytes) = tx.raw_get_bytes(&key.to_bytes()).await? {
+            value_bytes
+        } else {
+            struct CancelGaurd<F: FnOnce()>(Option<F>);
+            impl<F: FnOnce()> Drop for CancelGaurd<F> {
+                fn drop(&mut self) {
+                    self.0.take().unwrap()();
+                }
+            }
+            let notifications = self.notifications.clone();
+            // maybe use Arc instead?
+            let key_bytes2 = key_bytes.clone();
+            // cancellation behaviour: if one wait_key gets cancelled, all others with same key will
+            // never return, this is correct enough in case of timeouts
```

Comment:
I am not sure of better way of implementing this.
In practice, if one `wait_key` times out then another `wait_key` with same key that is being run concurrently will also timeout. New `wait_key` will just register again as normal.